sensor/tests/replay/replay_test.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/testutils/envisolator"
+	"github.com/stackrox/rox/pkg/utils"
+	centralDebug "github.com/stackrox/rox/sensor/debugger/central"
+	"github.com/stackrox/rox/sensor/debugger/k8s"
+	"github.com/stackrox/rox/sensor/debugger/message"
+	"github.com/stackrox/rox/sensor/kubernetes/listener/resources"
+	"github.com/stackrox/rox/sensor/kubernetes/sensor"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func TestReplayEvents(t *testing.T) {
+	suite.Run(t, new(ReplayEventsSuite))
+}
+
+type ReplayEventsSuite struct {
+	suite.Suite
+	fakeClient  *k8s.ClientSet
+	fakeCentral *centralDebug.FakeService
+	envIsolator *envisolator.EnvIsolator
+}
+
+var _ suite.SetupTestSuite = (*ReplayEventsSuite)(nil)
+var _ suite.TearDownTestSuite = (*ReplayEventsSuite)(nil)
+
+func (suite *ReplayEventsSuite) SetupTest() {
+	suite.fakeClient = k8s.MakeFakeClient()
+
+	suite.envIsolator = envisolator.NewEnvIsolator(suite.T())
+	suite.envIsolator.Setenv("ROX_MTLS_CERT_FILE", "../../../tools/local-sensor/certs/cert.pem")
+	suite.envIsolator.Setenv("ROX_MTLS_KEY_FILE", "../../../tools/local-sensor/certs/key.pem")
+	suite.envIsolator.Setenv("ROX_MTLS_CA_FILE", "../../../tools/local-sensor/certs/caCert.pem")
+	suite.envIsolator.Setenv("ROX_MTLS_CA_KEY_FILE", "../../../tools/local-sensor/certs/caKey.pem")
+
+	suite.fakeCentral = centralDebug.MakeFakeCentralWithInitialMessages(
+		message.SensorHello("1234"),
+		message.ClusterConfig(),
+		message.PolicySync([]*storage.Policy{}),
+		message.BaselineSync([]*storage.ProcessBaseline{}))
+}
+
+func (suite *ReplayEventsSuite) TearDownTest() {
+	suite.envIsolator.RestoreAll()
+}
+
+func createConnectionAndStartServer(fakeCentral *centralDebug.FakeService) (*grpc.ClientConn, *centralDebug.FakeService, func()) {
+	buffer := 1024 * 1024
+	listener := bufconn.Listen(buffer)
+
+	server := grpc.NewServer()
+	central.RegisterSensorServiceServer(server, fakeCentral)
+
+	go func() {
+		utils.IgnoreError(func() error {
+			return server.Serve(listener)
+		})
+	}()
+
+	conn, err := grpc.DialContext(context.Background(), "", grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+		return listener.Dial()
+	}), grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	if err != nil {
+		panic(err)
+	}
+
+	closeF := func() {
+		utils.IgnoreError(listener.Close)
+		server.Stop()
+	}
+
+	return conn, fakeCentral, closeF
+}
+
+var _ io.Writer = (*TraceWriterWithChannel)(nil)
+
+type TraceWriterWithChannel struct {
+	destinationChannel chan *central.SensorEvent
+	// mu mutex to avoid multiple goroutines writing at the same time
+	mu sync.Mutex
+}
+
+func (tw *TraceWriterWithChannel) Write(b []byte) (int, error) {
+	tw.mu.Lock()
+	defer tw.mu.Unlock()
```

Comment:
What is the shared resource that we protect with the mutex? The channel?

If yes, then I am not sure that we need it.